PhpDocs update

xtrime-ru opened the pull request:
1. Fixed style of generated phpDocs to allow IDE use them. 
2. Replaced callbacks with JsHandler
3. Detailed arguments in traits.

Tested in PHPStorm.

P.S. Docs generation from js code is awesome yet simple to read!

Commit:
Fix phpdoc format for IDE

Commit:
Add missing methods to Puppeteer class phpDocs

nesk commented on composer.json:
```diff
@@ -20,14 +20,16 @@
     "require": {
         "php": ">=7.3",
+        "ext-json": "*",
         "nesk/rialto": "^1.2.0",
         "psr/log": "^1.0",
         "vierbergenlars/php-semver": "^3.0.2"
     },
     "require-dev": {
         "monolog/monolog": "^2.0",
         "phpunit/phpunit": "^9.0",
-        "symfony/process": "^4.0|^5.0"
+        "symfony/process": "^4.0|^5.0",
+        "symfony/console": "^4.0|^5.0"
```
Good catch, those were missing. However, if you want to see this merged, I would rather want to have this in a separate commit. 🙂

xtrime-ru replied:
Sure. Will do in hour or two.

It was easy to find because docs generation don't work without it :)

xtrime-ru replied:
https://github.com/rialto-php/puphpeteer/pull/128

nesk commented on src/Puppeteer.php:
```diff
@@ -10,11 +12,16 @@
 /**
  * @property-read mixed devices
  * @property-read mixed errors
- * @method \Nesk\Puphpeteer\Resources\Browser connect(array<string, mixed> $options)
+ * @property-read string product
+ * @method Browser connect(array $options)
+ * @method Browser launch(array $options = [])
+ * @method string[] defaultArgs(array $options = [])
  * @method void registerCustomQueryHandler(string $name, mixed $queryHandler)
  * @method void unregisterCustomQueryHandler(string $name)
  * @method string[] customQueryHandlerNames()
  * @method void clearCustomQueryHandlers()
+ * @method BrowserFetcher createBrowserFetcher(array $options = [])
+ * @method string executablePath()
```
Did you add this manually? I can't find what in `src/doc-generator.ts` can alter this?

xtrime-ru replied:
Ah, probably yes and forgot to mention.

xtrime-ru replied:
I fixed \Nesk\Puphpeteer\Command\GenerateDocumentationCommand::execute to add missing methods and properties from PuppeteerNode class.

xtrime-ru commented:
Hello! 

Can anybody check this PR and merge or send feedback?

nesk commented:
Hi! I've started to write a review, I will try to finish it tonight.

xtrime-ru commented:
Many thanks! There is no hurry. I just wanted to know, if i need to wait little bit more or this PR will be buried in sands of time :)

xtrime-ru commented:
To make review simpler, here is the list of files i have changed:

- composer.json 
- src/Command/GenerateDocumentationCommand.php
- src/Traits/AliasesEvaluationMethods.php
- src/Traits/AliasesSelectionMethods.php
- src/doc-generator.ts


After that i run docs generator cli command to update automatically generated phpDocs.

nesk commented on src/Resources/Browser.php:
```diff
@@ -4,15 +4,15 @@
- * @method \Nesk\Puphpeteer\Resources\BrowserContext createIncognitoBrowserContext()
- * @method \Nesk\Puphpeteer\Resources\BrowserContext[] browserContexts()
- * @method \Nesk\Puphpeteer\Resources\BrowserContext defaultBrowserContext()
+ * @method BrowserContext createIncognitoBrowserContext()
+ * @method BrowserContext[] browserContexts()
+ * @method BrowserContext defaultBrowserContext()
```
I'm not sure about this, by using FQCN references we are sure we will never have anything breaking due to wrong relative references. Why did you choose to do this?

xtrime-ru replied:
I alsways use simple FQCN, because its much easier to read. And i think IDE add relative class names if its same namespace by default. 

Problems can occure if some classes will move to other folder. But i dont think it will happen. 

Anyway prefix controlled by one string in console command. And if there will be different namespaces it will cause troubles too. So i suggest to solve problems only if they arrive.

nesk replied:
> Anyway prefix controlled by one string in console command. And if there will be different namespaces it will cause troubles too. So i suggest to solve problems only if they arrive.

This is a really good point, but I would agree only if the classes references weren't messed in `src/Puppeteer.php`:

https://github.com/rialto-php/puphpeteer/blob/03ee95752181c948c5b3f4dedbfcdcfbe082bd5e/src/Puppeteer.php#L16-L17

This is not valid, the proper references should be:

```diff
-  * @method Browser connect(array $options) 
-  * @method Browser launch(array $options = []) 
+  * @method Resources\Browser connect(array $options) 
+  * @method Resources\Browser launch(array $options = []) 
```

By using FQCN references we are sure we will never have any issues like that. I think changing a single value in the command is not something really problematic compared to potential failures in relative class references.

xtrime-ru replied:
> https://github.com/rialto-php/puphpeteer/blob/03ee95752181c948c5b3f4dedbfcdcfbe082bd5e/src/Puppeteer.php#L16-L17
> 
> This is not valid, the proper references should be:
> 
> ```diff
> -  * @method Browser connect(array $options) 
> -  * @method Browser launch(array $options = []) 
> +  * @method Resources\Browser connect(array $options) 
> +  * @method Resources\Browser launch(array $options = []) 
> ```
> 

This is valid, because we have: 
https://github.com/rialto-php/puphpeteer/blob/03ee95752181c948c5b3f4dedbfcdcfbe082bd5e/src/Puppeteer.php#L5

nesk commented on src/Resources/Browser.php:
```diff
@@ -4,15 +4,15 @@
+ * @method Page newPage()
+ * @method Target[] targets()
+ * @method Target target()
+ * @method Target waitForTarget(JSHandle $predicate, array $options = [])
```
This seems wrong, [`waitForTarget` only accepts a callback for its first parameter](https://pptr.dev/#?product=Puppeteer&version=v5.5.0&show=api-browserwaitfortargetpredicate-options).

nesk replied:
I'm re-thinking about that, maybe you wanted to use `Nesk\Rialto\Data\JsFunction` instead?

xtrime-ru replied:
Exactly! Cant find this class now :( 
UPDATE: 
Nevermind. I just realised this is class from another package... :)

nesk commented on src/Resources/Coverage.php:
```diff
@@ -5,9 +5,9 @@
 /**
- * @method void startJSCoverage(array<string, mixed> $options = null)
+ * @method void startJSCoverage(array $options = [])
  * @method mixed[] stopJSCoverage()
- * @method void startCSSCoverage(array<string, mixed> $options = null)
+ * @method void startCSSCoverage(array $options = [])
```
Actually, those types are perfectly valid and I'm not the only one using them. This is supported by [PHPStan](https://phpstan.org/writing-php-code/phpdoc-types#general-arrays) and [Psalm](https://psalm.dev/docs/annotating_code/type_syntax/array_types/#generic-arrays).

Also, [PHPStorm should support those syntaxes](https://blog.jetbrains.com/phpstorm/2020/07/phpstan-and-psalm-support-coming-to-phpstorm/), maybe you didn't update to the latest version?

xtrime-ru replied:
![Снимок экрана 2021-02-09 в 22 20 07](https://user-images.githubusercontent.com/34161928/107415957-f8acc900-6b24-11eb-8221-253ce58100b7.png)

Im using latest available version. In your link such types support implemented as plugins. Probably something went wrong with devs plans and they were not added/enabled by default...

Also there can be problems with other IDEs, since its not standart php syntax. 

Maybe we can use php 8 annotations? But i did not used them before.

xtrime-ru replied:
I just checked release notes https://blog.jetbrains.com/phpstorm/2020/12/phpstorm-2020-3-release/#psalm_and_phpstan_support 

Such types can be used in project if Psalm or PHPStan added in composer and inspections enabled in settings. 

Also custom attributes added to phpStorm, but i doubt they will not be supported in different IDE's https://blog.jetbrains.com/phpstorm/2020/12/phpstorm-2020-3-release/#custom_php_attributes_in_phpstorm . So its not best idea to use #[ArrayShape].

I will try to add psalm tom composer and setup my IDE. Maybe most of my edits will be not needed after that.

nesk replied:
Not all IDE's may support these annotations today, but it's really helping if you properly configure your CI with a static analysis tool like Psalm, PHPStan, etc. And support will improve in the next months/years. 🙂

For example, recently I was using [getsentry/sentry-php](https://github.com/getsentry/sentry-php) package, and their extended annotations saved me multiple times.

xtrime-ru replied:
I checked phpStan and Psalm support in phpStorm, and they both don't support such syntax in `@method` block
![Снимок экрана 2021-02-16 в 18 08 17](https://user-images.githubusercontent.com/34161928/108081374-f3e69880-7081-11eb-8105-cac84a782903.png)

But same syntax in `@param` or `@return` blocks works ok (added phpDoc only for tests):
![Снимок экрана 2021-02-16 в 18 18 42](https://user-images.githubusercontent.com/34161928/108082686-6ad06100-7083-11eb-8cc2-789c2993cdff.png)

So my plan is: 

1. add phpStan to composer dev, to enable as much phpDocs syntax in IDE as possible
2. Revert changes in traits
3. Fix bugs with JsHandle/JsFunction
4. Fix missing methods in Puppeteer class (maybe exclude it from docs generation?)

What do you think? Do you insist on full namespaces instead of FQCN?

nesk commented on src/Traits/AliasesSelectionMethods.php:
```diff
@@ -2,25 +2,30 @@
 
 namespace Nesk\Puphpeteer\Traits;
 
-/**
- * @method ElementHandle|null querySelector(string $selector)
- * @method array<int, ElementHandle> querySelectorAll(string $selector)
- * @method array<int, ElementHandle> querySelectorXPath(string $expression)
- */
```
I'm not in favor of removing the doc blocks for native typings. The reason is because we can easily support a new Puppeteer version without having to edit the code.

xtrime-ru replied:
Hm. Good point, but:

1. this file don't covered by generator. 
2. i think all puppeteer versions have same syntax for this. And i think it will remain. Its very basic actions.

nesk replied:
>     1. this file don't covered by generator.

Oh! That's right, my bad. 😅 

>     2. i think all puppeteer versions have same syntax for this. And i think it will remain. Its very basic actions.

Actually, we can't really know or be sure about that. They could perfectly add a second parameter to set options like extended filtering. I know this is not very likely to happen, but the goal here with PuPHPeteer is to act **as a simple proxy** and no more. If something new is added then I want to support it without having to update the code, otherwise it would require more maintainance (new issues, new PRs, new releases, etc.).

But maybe I'm missing something? Does using phpDocs **blocks** you with the way you use the PuPHPeteer? If so, explain me so I can understand and either provide you a solution or approve those changes.

xtrime-ru replied:
> But maybe I'm missing something? Does using phpDocs **blocks** you with the way you use the PuPHPeteer? If so, explain me so I can understand and either provide you a solution or approve those changes.

Well, its not blocking, but psalm syntax was not supported by my IDE. So there is no suggests and syntax checks. But if i manage to setup PHPStan to parse such phpDocs format in realtime then most of changes can be reverted. Its new feature in phpStorm

nesk commented:
>P.S. Docs generation from js code is awesome yet simple to read!

I'm glad it helps you!

xtrime-ru committed:
Fix Traits Docs

xtrime-ru committed:
Add PhpStan for IDE support

xtrime-ru committed:
Add Puppeteer missing methods

xtrime-ru committed:
Remove build js before docs generation.

xtrime-ru committed:
JsFunction type instead of callbacks

xtrime-ru commented on composer.json:
```diff
@@ -28,7 +29,8 @@
         "symfony/process": "^4.0|^5.0",
-        "symfony/console": "^4.0|^5.0"
+        "symfony/console": "^4.0|^5.0",
+        "phpstan/phpstan": "^0.12"
```
Its not necessary, because "rich types" not supported for virtual methods...

xtrime-ru commented:
@nesk Ive reverted Traits changes. 
But i still think we need to fix them :) : 
![image](https://user-images.githubusercontent.com/34161928/108222877-3f647980-714a-11eb-83f5-657988f02e8e.png)

Also i've done some stuff: 
1. Add Puppeteer methods from Puppeteer and PuppeteerNode
2. Remove build folder before building docs
3. Added JsFunction type instead of callbacks. Also added replacement of mixed to JsFunction for evaluation* methods.

Will wait for your approval and merge ;)

nesk commented:
# IDE autocompletion issues

I have downloaded PHPStorm and tried to setup everything, I must agree: it doesn't work well. 😅 

However, you can clearly see [with this PHPStan demo](https://phpstan.org/r/1d14ce4d-9458-4a2a-9be6-7d722697ca0d) what it brings to have those annotations, they really help. I'm don't want to give up on them, but I have some ideas to fix the current issues.

Here are some interesting points:

- PHPStan works really well with the documentation, as seen above.
- [Psalm doesn't work with the way the `@method` annotations are written](https://psalm.dev/r/c14d3b58a2), I wasn't expecting that.
- PHPStorm doesn't work with the documentation, as you said.

The good news:

- PHPStan can use `@phpstan-method` instead of `@method`, which would allow to fully type the methods without breaking Psalm and IDEs.
- Psalm can use `@psalm-method` instead of `@method`, which would allow to use better typings than the ones for IDEs and not break on PHPStan format.
- In addition, we can add simple `@method` annotations which are working everywhere and will allow IDEs to have proper autocompletion.

By using the three documentation formats here, we could achieve perfect support everywhere.

# Callback documentation

You are right, the current typings for the callbacks aren't the good ones, because PuPHPeteer expects you to use a `JsFunction` instance. You've made the following changes:

```diff
- * @method mixed download(string $revision, callable(float $x, float $y): void $progressCallback = null)
+ * @method mixed download(string $revision, \Nesk\Rialto\Data\JsFunction $progressCallback = null)
```

This is correct, however we lose the detailed typings. I think we can fix this by doing this:

```diff
- * @method mixed download(string $revision, callable(float $x, float $y): void $progressCallback = null)
+ * @method mixed download(string $revision, (callable(float, float): void)|JsFunction $progressCallback = null)
```

This works perfectly and can still help users to see the detailed typings, [here's an example](https://phpstan.org/r/9c3d21fe-96d3-42ab-95f0-ee74b5453d02).

Additionally, I could add in Rialto a trigger to warn users about the fact they are using a `callable` instead of a `JsFunction` instance.

# FQCN references

> What do you think? Do you insist on full namespaces instead of FQCN?

I've tought about it, I see no difference on your side (as a user) if I use FQCN refs or relative ones. As the maintainer of the project, I would rather keep the FQCN references since I find them more reliable.

PHPStorm makes no difference between relative or FQCN references, it always displays the FQCN one:

<img width="380" alt="Capture d’écran 2021-02-18 à 13 36 03" src="https://user-images.githubusercontent.com/817508/108358063-8c9a2700-71ee-11eb-9a5a-c530152c7579.png" align="left">
<img width="380" alt="Capture d’écran 2021-02-18 à 13 36 15" src="https://user-images.githubusercontent.com/817508/108358068-8d32bd80-71ee-11eb-9c29-fc8721785ab1.png" align="left">

# To conclude

I see immediate benefits in some of your changes:

- You have fixed the missing `symfony/console` dependency. This is already merged in another PR, thank you again!
- You have fixed the missing `$` character in `src/Traits/AliasesEvaluationMethods.php`.
- You have fixed the missing properties and methods in `src/Puppeteer.php`.

If you want, you can make a new PR with only the last two changes in this list and I will merge it right away. 🙂

About everything else, there is still work to do:

- The documentation generator should output three formats for each property/method/etc to support the majority of the tools, namely PHPStan, Psalm and all other tools reading phpDoc (like IDEs).
- The documentation for callbacks should support `JsFunction` instances without removing the detailed typings, a union can solve this.

If you don't have time anymore to work on these, I can take over. However, I may not have time to work on this soon so you will have to deal with the current typings for a while.

xtrime-ru commented:
1. >IDE autocompletion issues

Good idea to use `@method`, `@phpstan-method`, `@psalm-method`. I will modify doc-genereator. Add some property, to contriol output format of docs. We will run it 3 times, with different formats, and then merge them and write to phpDocs.

2. >Callback documentation 

Union is ok, i will do that.

3. >FQCN references

Ok. Will return full namespace


Thanks for nice ideas, will implement them in upcoming days.

Commit:
Multiple phpDoc formats: php and phpstan

xtrime-ru commented:
Hi! Sorry for delay with this fixes. 

Fixed: FQCN references and callbacks unions (for phpstan format).
Added support for multiple doc styles for php.

But I ve run into another issue: phpStorm don't support @phpstan-method with our annotations. And it complains about duplicate methods: 
![image](https://user-images.githubusercontent.com/34161928/114751405-cbb3a700-9d5d-11eb-9f87-9d669e10ce08.png)

also hints are incorect: 
![image](https://user-images.githubusercontent.com/34161928/114751570-fb62af00-9d5d-11eb-8dbb-0ed18886db74.png)

My proposal: replace `@phpstan-method` with non annotation string, like `@method-extended`. This is only way i can manage phpStorm works:
![image](https://user-images.githubusercontent.com/34161928/114751920-6f04bc00-9d5e-11eb-8d78-4dacfa9279a9.png)

xtrime-ru commented:
@nesk Hi! Want to remind about my pull request :)

xtrime-ru commented:
Hi! PhpStorm got update with generics support in PHPDocs. Maybe this will help to cleanup this PR.

xtrime-ru commented:
>Hi! PhpStorm got update with generics support in PHPDocs. Maybe this will help to cleanup this PR.

Checked in new PhpStorm and there is still no support for array shape for argement types in `@method`. So PR will be same.

xtrime-ru committed:
Fix: Puppeteer 10. Basic support arrow functions.

Commit:
Puppeteer 14 docs support.
Arrow functions fix & #private tags
